## Section 1 + 2
***please see requirements.txt***

##  Section 3 + 4
## loads data and Compute simple statistics on the data

In [1]:
from analysis import load_data, analyze_spam_csv

In [2]:
# File path to the CSV file
file_path = "data/spam.csv"

data = load_data(file_path)

# Analyze and print statistics
stats = analyze_spam_csv(data)
for key, value in stats.items():
    print(f"{key}: {value}")

Number of SMS messages: 5572
Number of spam messages: 747
Total word count: 86835
Average words per message: 15.584
5 most frequent words: [('to', 2248), ('i', 2239), ('you', 2128), ('a', 1442), ('the', 1330)]
Number of rare words: 5276


## Section 5 Perform tokenization

In [3]:
from tokenize_nltk import tokenize_nltk

[nltk_data] Downloading package punkt to /Users/ranwright/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [4]:
# Tokenize using NLTK
nltk_tokens = tokenize_nltk(data["Message"])
print(f"Total NLTK tokens: {len(nltk_tokens)}")

Total NLTK tokens: 104904


In [5]:
from tokenize_spacy import tokenize_spacy

In [6]:
# Tokenize using spaCy
spacy_tokens = tokenize_spacy(data["Message"])
print(f"Total spaCy tokens: {len(spacy_tokens)}")

Total spaCy tokens: 104367


In [7]:
from analysis import compare_tokenizations

In [8]:
compare_token_method_result = compare_tokenizations(nltk_tokens, spacy_tokens)
for key, value in compare_token_method_result.items():
    print(f"{key}: {value}")

NLTK unique tokens count: 689
spaCy unique tokens count: 743
Common tokens count: 10822
Total NLTK tokens: 104904
Total spaCy tokens: 104367


## Section 6 Perform lemmatization

In [9]:
from lemmatization_nltk import lemmatize_nltk

[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/ranwright/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     /Users/ranwright/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/ranwright/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /Users/ranwright/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!


In [10]:
# Lemmatize using NLTK
nltk_lemmas = lemmatize_nltk(data["Message"])
print(f"Total NLTK lemmas: {len(nltk_lemmas)}")

Total NLTK lemmas: 104904


In [11]:
from lemmatization_spacy import lemmatize_spacy

In [12]:
# Lemmatize using spaCy
spacy_lemmas = lemmatize_spacy(data["Message"])
print(f"Total spaCy lemmas: {len(spacy_lemmas)}")

Total spaCy lemmas: 104367


In [13]:
from analysis import compare_lemmatizations

In [14]:
# Compare lemmatizations
lemma_diff = compare_lemmatizations(nltk_lemmas, spacy_lemmas)
print("\nComparison of Lemmatizations:")
for key, value in lemma_diff.items():
    print(f"{key}: {value}")


Comparison of Lemmatizations:
NLTK unique lemmas count: 2379
spaCy unique lemmas count: 1434
Common lemmas count: 8221
Total NLTK lemmas: 104904
Total spaCy lemmas: 104367


## Section 7 Perform stemming

In [15]:
from stem_nltk import stem_nltk

[nltk_data] Downloading package punkt to /Users/ranwright/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [16]:
# Stem using NLTK (Porter Stemmer)
nltk_stems = stem_nltk(data["Message"], method="porter")
print(f"Total NLTK (Porter) stems: {len(nltk_stems)}")

Total NLTK (Porter) stems: 104904


In [17]:
from stem_spacy import stem_spacy

In [18]:
# Stem using spaCy (approximation)
spacy_stems = stem_spacy(data["Message"])
print(f"Total spaCy stems: {len(spacy_stems)}")

Total spaCy stems: 104367


In [19]:
from analysis import compare_stemmings

In [20]:
# Compare stemmings
stem_diff = compare_stemmings(nltk_stems, spacy_stems)
print("\nComparison of Stemmings:")
for key, value in stem_diff.items():
    print(f"{key}: {value}")


Comparison of Stemmings:
NLTK unique stems count: 5870
spaCy unique stems count: 4777
Common stems count: 2197
Total NLTK stems: 104904
Total spaCy stems: 104367


## Section 8 Finding a Message That Reduces Stems but Keeps Lemmas Unchanged

This analysis identifies a message from the dataset where removing it would:
- Reduce the total number of unique **stemmed tokens**.
- Keep the total number of unique **lemmatized tokens** unchanged.

The result highlights cases where stemming is more aggressive and fragmented compared to lemmatization, which tends to normalize words consistently to their base forms.


In [21]:
from comparison import find_message_with_stem_lemma_disparity

In [22]:
# Example usage
result_message = find_message_with_stem_lemma_disparity(
    messages=data["Message"].tolist(), nltk_stems=nltk_stems, nltk_lemmas=nltk_lemmas
)

if result_message:
    print("Message causing the effect:\n")
    print(
        repr(result_message)
    )  # Ensures the entire message, including multiline, is printed
else:
    print("No message found that meets the criteria.")

Message causing the effect:

'Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...'


## Section 9 Finding a Message That Reduces Lemmas but Keeps Stems Unchanged

This analysis identifies a message from the dataset where removing it would:
- Reduce the total number of unique **lemmatized tokens**.
- Keep the total number of unique **stemmed tokens** unchanged.

The result highlights cases where stemming and lemmatization handle words differently, emphasizing the linguistic nuances captured by lemmatization but overlooked by stemming.


In [23]:
from comparison import find_message_with_lemma_stem_disparity

In [24]:
# Example usage
result_message = find_message_with_lemma_stem_disparity(
    messages=data["Message"].tolist(), nltk_stems=nltk_stems, nltk_lemmas=nltk_lemmas
)

if result_message:
    print("Message causing the effect:\n")
    print(
        repr(result_message)
    )  # Ensures the entire message, including multiline, is printed
else:
    print("No message found that meets the criteria.")

Message causing the effect:

'Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...'


<br>----------------------------------------------<br>
## Section 10. Comparison of the results NLTK vs. SpaCy

#### Tokenization:
- **NLTK Unique Tokens**: 689
- **SpaCy Unique Tokens**: 743
- **Common Tokens**: 10,822
- **Total Tokens**:
  - NLTK: 104,904
  - SpaCy: 104,367

**Observation**:
- SpaCy identifies slightly more unique tokens due to its advanced handling of special characters, contractions, and linguistic nuances.

#### Lemmatization:
- **NLTK Unique Lemmas**: 2,379
- **SpaCy Unique Lemmas**: 1,434
- **Common Lemmas**: 8,221
- **Total Lemmas**:
  - NLTK: 104,904
  - SpaCy: 104,367

**Observation**:
- SpaCy's lemmatization is more conservative, producing fewer unique lemmas, while NLTK is more aggressive and often generates more variations.

#### Stemming:
- **NLTK Unique Stems**: 5,870
- **SpaCy Unique Stems**: 4,777
- **Common Stems**: 2,197
- **Total Stems**:
  - NLTK: 104,904
  - SpaCy: 104,367

**Observation**:
- NLTK's stemming is more aggressive, leading to more fragmented stems. SpaCy’s stems (approximated using `.lemma_`) are closer to lemmatization and thus less varied.


<br>----------------------------------------------<br>
## Section 11. Comparison of NLTK Before and After Tokenization, Lemmatization, and Stemming

#### Before Tokenization:
- Messages are raw text with varying formats (e.g., punctuation, special characters, etc.).

#### After Tokenization:
- **Total Tokens**: 104,904
- **Unique Tokens**: 689
- NLTK tokenization splits text into words but does not handle linguistic nuances (e.g., contractions are split into separate tokens).

#### After Lemmatization:
- **Total Lemmas**: 104,904
- **Unique Lemmas**: 2,379
- Lemmatization reduces tokens to their dictionary forms but keeps linguistic variations (e.g., "better" → "good").

#### After Stemming:
- **Total Stems**: 104,904
- **Unique Stems**: 5,870
- Stemming is more aggressive, often stripping suffixes and creating non-dictionary words (e.g., "running" → "run", "better" → "better").

**Observation**:
- Tokenization introduces a structured list of words.
- Lemmatization normalizes words to base forms, leading to fewer unique entries.
- Stemming is more aggressive and results in more unique variations than lemmatization.


<br>-----------------------------------------<br>
## section 12. Comparison of SpaCy Before and After Tokenization, Lemmatization, and Stemming

#### Before Tokenization:
- Messages are raw text, similar to the NLTK input.

#### After Tokenization:
- **Total Tokens**: 104,367
- **Unique Tokens**: 743
- SpaCy handles linguistic details better, preserving contractions (e.g., "I'm" remains a single token).

#### After Lemmatization:
- **Total Lemmas**: 104,367
- **Unique Lemmas**: 1,434
- SpaCy produces fewer unique lemmas, as it normalizes words more consistently to their dictionary forms (e.g., "better" → "good", "running" → "run").

#### After Stemming (Approximated):
- **Total Stems**: 104,367
- **Unique Stems**: 4,777
- SpaCy's stemming (approximated using `.lemma_[:n]`) is closer to its lemmatization, resulting in fewer unique variations.

**Observation**:
- Tokenization splits text more precisely, maintaining contractions as single tokens.
- Lemmatization reduces redundancy by normalizing words effectively.
- Stemming (if approximated) produces results closer to lemmatization compared to NLTK.
